src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/UnspeakableTypeTests.cs

```diff
@@ -0,0 +1,154 @@
+        [Theory]
+        [MemberData(nameof(GetUnspeakableTypes))]
+        public async Task CanSerializeUnspeakableTypesAsBoxedProperties<T>(Envelope<T> envelope, string expectedJson)
+        {
+            var boxedEnvelope = new Envelope<object>(envelope.Value);
+            string expectedEnvelopeJson = $$"""{"Value":{{expectedJson}}}""";
+
+            string json = await Serializer.SerializeWrapper(boxedEnvelope, UnspeakableTypeContext.Default.Options);
+            Assert.Equal(expectedEnvelopeJson, json);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetUnspeakableTypes))]
+        public async Task CanSerializeUnspeakableTypesAsBoxedCollectionElements<T>(Envelope<T> envelope, string expectedJson)
+        {
+            var boxedEnvelope = new List<object> { envelope.Value };
+            string expectedCollectionJson = $"[{expectedJson}]";
+
+            string json = await Serializer.SerializeWrapper(boxedEnvelope, UnspeakableTypeContext.Default.Options);
+            Assert.Equal(expectedCollectionJson, json);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetUnspeakableTypes))]
+        public async Task CanSerializeUnspeakableTypesAsBoxedDictionaryValues<T>(Envelope<T> envelope, string expectedJson)
+        {
+            var boxedEnvelope = new Dictionary<string, object> { ["key"] = envelope.Value };
+            string expectedCollectionJson = $$"""{"key":{{expectedJson}}}""";
+
+            string json = await Serializer.SerializeWrapper(boxedEnvelope, UnspeakableTypeContext.Default.Options);
+            Assert.Equal(expectedCollectionJson, json);
+        }
+
+        [Fact]
+        public async Task TypeWithDiamondAmbiguityThrowsNotSupportedException()
+        {
+            object value = new TypeWithDiamondAmbiguity();
+
+            NotSupportedException exn = await Assert.ThrowsAsync<NotSupportedException>(() => Serializer.SerializeWrapper(value, UnspeakableTypeContext.Default.Options));
+            Assert.Contains("TypeWithDiamondAmbiguity", exn.Message);
+            Assert.Contains("BasePoco", exn.Message);
+            Assert.Contains("IEnumerable`1[System.Int32]", exn.Message);
+
+            exn = await Assert.ThrowsAsync<NotSupportedException>(() => Serializer.SerializeWrapper(value, value.GetType(), UnspeakableTypeContext.Default.Options));
+            Assert.Contains("TypeWithDiamondAmbiguity", exn.Message);
+            Assert.Contains("BasePoco", exn.Message);
+            Assert.Contains("IEnumerable`1[System.Int32]", exn.Message);
+        }
+
+        public static IEnumerable<object[]> GetUnspeakableTypes()
+        {
+            yield return Wrap<IEnumerable<int>>(new int[] { 1, 2, 3 }, "[1,2,3]");
+            yield return Wrap(Enumerable.Range(1, 3), "[1,2,3]");
+            yield return Wrap(Enumerable.Range(1, 3).Select(x => x - 1), "[0,1,2]");
+            yield return Wrap(IteratorMethod(), "[1,2,3]");
+            yield return Wrap(AsyncIteratorMethod(), "[1,2,3]");
```

Comment:
You've mentioned private interface implementations a few times in your responses to folks.  Do any of these tests exercise that?

Reply:
I expect these to be covered by the `IEnumerable` and `IAsyncEnumerable` test inputs.

Reply:
What properties are implemented explicitly in those cases?  Or maybe I misunderstood what case you were referring to?

Reply:
If you mean POCO-shaped interfaces, then no, I've skipped that case. The fallback logic is shape-invariant and it's complicated by the fact that interfaces don't support roundtripping OOTB. I'll add a separate test covering that just in case.